src/algebra/associated.lean

```diff
@@ -135,12 +136,16 @@ hp.1
 
 lemma is_unit_or_is_unit [monoid α] {p : α} (hp : irreducible p) {a b : α} (h : p = a * b) :
   is_unit a ∨ is_unit b :=
-hp.2 a b h
+irreducible.is_unit_or_is_unit' a b h
 
 end irreducible
 
+lemma irreducible_iff [monoid α] {p : α} :
```

Comment:
Is it worth it to try to generate these iff statements automatically (eg with an attribute)?

Reply:
I think so; ergonomics on these definitions has decreased quite a bit and there is a need to restate all the projections as lemmas, so getting some of that back with metaprogramming would be nice. In particular, making a definition a class for occasional use in `[]` even if it is normally used as a hypothesis is now not a "zero cost" operation, you have to box up the predicate and everything that uses it has to keep this in mind.

Reply:
I thought `mk_inductive_of_iff` was supposed to do this, but it failed with a crappy error message when I tried it and I didn't investigate.

Reply:
Thanks for pushing through with this!

> ergonomics on these definitions has decreased quite a bit

Looking at the diff I definitely agree that this is a regression and we should try to compensate somehow. Some of it looks automatable with little effort, some looks more subtle. Sorry this has been a crazy week and I haven't had time to think about it harder.

Reply:
Of course this is going to be a regression, we're straight up removing a lean 3 feature. The `@[class] def` construct is weird enough that it's usually only used where it is helpful, and so working around that means things get worse. My thoughts are to get this merged ASAP (it touches a lot of files so it bitrots quickly) and look for ways to improve ergonomics and tooling around this construct in a future PR.

Reply:
That's fair. I didn't mean to suggest holding up this PR. But this is something we need to think about in Lean 4, and possibly in Lean 3 if we want a "clean" final Lean 3 mathlib.